### ISSUE TITLE
Question about behavior when input is in CoNNL-U format

### ISSUE THREAD

**Issue by harisont** [ufal/udpipe#158]
Hi!

I'm trying to use UDPipe (version 2) on __almost complete CoNNL-U files__, where just a couple dependency labels, POS tags and/or lemmas are missing from each sentence.

My intuition was that, in cases like this, UDPipe would just "fill the gaps", but it seems that it actually also overwrites information that was present in the input file.

I wonder:
- is this intended?
- is there a way to use UDPipe that prevents this from happening, i.e. to just have it fill the gaps?

**Comment by martinpopel** [comment]
I think it is intended - when I run a parser, I expect it to parse the whole sentence (and ignore any dependencies on the input, if present). I am not aware of any way how to instruct UDPipe to use existing annotation and just "fill the gaps" - it would require a completely new parsing/tagging algorithm.

That said, it should be easy to write a script (e.g. using [Udapi](https://github.com/udapi/udapi-python)), which takes two conllu files with the same tokenization on the input and merges them, i.e. fills the "gaps" in the first one with annotation in the second one. Of course, one needs to make sure no cycles are introduced this way in the basic dependencies, but that is also easy. It may not give the optimal results (because UDPipe didn't use the pre-existing annotation), but it should be usable.

**Comment by foxik** [comment]
Hi all,

UDPipe just overwrites all the given annotations for a given column if it is to predict it. However, UDPipe currently does not benefit from partial annotation (i.e., knowing several POS tags when filling the others), so if you just want to "fill the gaps", you  are no worse in just copying the information you want to keep.

What I mean is that even if UDPipe knew some of the annotations are fixed, if would give you the same predictions anyway (maybe with the exception of dependency edges, where the MST algorithm might give a different result). But I understand you might just not want to do the copying yourself, and would prefer if UDPipe had an option of not overwriting existing annotations. Once UDPipe 3 is finished, I can easily add such a functionality, if there is demand.

(Note that we often call predict on the test sets (including the gold annotations), so ignoring the annotations sometimes is an intended behaviour.)

Cheers!

BTW, we often call predict on the test sets (including the gold annotations), so ignoring the annotations sometimes is an intended behaviour.

BTW2, surprisingly, if you are interested in parsing, then having access to gold POS tags does not make parsing better, at least not on Czech_PDT, where we performed the experiment...

**Comment by foxik** [comment]
> > UDPipe currently does not benefit from partial annotation
> 
> "Currently", so, out of curiosity, are there any plans to change this in future versions?

No active plans, no.

> > BTW2, surprisingly, if you are interested in parsing, then having access to gold POS tags does not make parsing better, at least not on Czech_PDT, where we performed the experiment...
> 
> Interesting! But I guess having the lemmas helps getting the right POS tags, and having the heads helps with dependency labels?

Yes, I would think so.

**Comment by jowagner** [comment]
I see two different workarounds:

1. Use a voting combiner such as https://github.com/jowagner/ud-combination with the following 3 inputs at weights 0.3, 0.3 and 0.4. When the annotation is not known yet, the third input wins 0.4 against 0.3. Otherwise, the existing label is preferred (0.6 against 0.4). Note though that the combiner guarantees a spanning tree and if necessary will deviate from some or all of the known labels. The dummy labels `NONEXISTINGLABEL1` and `NONEXISTINGLABEL2` should never appear in the combiner output as they share the same parent with the prediction and the label of the prediction has higher weight. 
    1. The existing annotation with empty fields populated with the arc `(NONEXISTINGLABEL1, x)` where `x` is the head position in UDPipe's prediction
    2. The existing annotation with empty fields populated with the arc `(NONEXISTINGLABEL2, x)`
    3. The prediction of UDPipe

2. UDPipe 2 supports reading an external vector representation for each input token. Normally this is used to load context-sensitive word embeddings such as Elmo and BERT (frozen, no fine-tuning). One could use this input channel to provide the neural model with arbitrary per-token inputs such as the existing dependency annotations discussed here. Of course, to use such new inputs one must train a new model and the training must be done in a way that the model learns both making use of the encoded known dependency relations and predicting new dependency relations were necessary. Also, the encoding of the dependency labels and heads may need some attention. A simple one-hot encoding may work for the labels but the heads will need some kind of position encoding. It should be noted though that this approach does not guarantee all know labels and heads will be copied, only that the extra information will be used to the extent that it proved useful in training.
